### ISSUE TITLE
Faking the Referer Header in PhantomJS

### ISSUE THREAD

**Issue by lonalore** [serp-spider/http-client-phantomjs#4]
Hi,

How can I set a fake "Referer" value for PhantomJS? Is it possible?

**My code snippet:**
```php
$client = new PhantomJsClient();

$userAgent = UserAgent::random([
  'os_type'     => 'Windows',
  'device_type' => 'Desktop',
]);

$proxy = new Proxy(PROXY_HOST, PROXY_PORT, PROXY_USER, PROXY_PASS);

$browser = new Browser($client, $userAgent, SEARCH_LANGUAGE);
$browser->setProxy($proxy);

$googleClient = new GoogleClient($browser);

$googleUrl = new GoogleUrl(SEARCH_GOOGLE_HOST);
$googleUrl->setSearchTerm(SEARCH_KEYWORD);
$googleUrl->setPage(1);

$response = $googleClient->query($googleUrl);
```

Thank you!

**Comment by gsouf** [comment]
Hi @lonalore 

At this moment this is something impossible. But we can work on this feature.

**Comment by lonalore** [comment]
Well, I made some modifications for achieving this. Now I'm able to set my custom referer value as follows:

```php
$client = new PhantomJsClient();
$client->setPhantomJsHeader('Referer', 'https://www.google.hu/');
```

You can check my changes here: https://github.com/serp-spider/http-client-phantomjs/compare/master...lonalore:custom-headers

If it suits you, I could make a pull-request.

**Comment by lonalore** [comment]
Or any other solution we can implement? :) I'm open to everything. :)

**Comment by gsouf** [comment]
Hi @lonalore, sorry for the delay.

Thanks for your proposition. Your solution is interesting, and I think I will merge it.

However it is limited to phantomjs and we can have a more comon way to solve the problem.
Actually when we send a request to google, we define the request on php side (that means it can be common to curl + phantomjs + future adapters like headless chrome). The referer  is only an http header, and I"m thinking of a solution where you would add http headers on the request from the google client before it is passed to the http client.

For your solution feel free to propose the PR, I will review as soon as possible!

By the way, thanks for contributing, it's appreciated!

**Comment by lonalore** [comment]
Hi! Thank you for your answer. I made some modifications according to your comment. With my changes, we could use **setHeader()** and **getHeaders()** methods in curl + phantomjs + future adapters. But for this, I had to change files in the following repositories:

**HttpClientInterface.php** in **core** repository
https://github.com/serp-spider/core/compare/master...lonalore:ch

**CurlClient.php** in **http-client-curl** repository
https://github.com/serp-spider/http-client-curl/compare/master...lonalore:ch

**PhantomJsClient.php** in **http-client-phantomjs** repository
https://github.com/serp-spider/http-client-phantomjs/compare/master...lonalore:ch

**GoogleClient.php** in **search-engine-google** repository
https://github.com/serp-spider/search-engine-google/compare/master...lonalore:ch


**Example usage:**

```php
if (CLIENT_TYPE == CLIENT_TYPE_PHANTOMJS) {
  $client = new PhantomJsClient();
} else {
  $client = new CurlClient();
}

$userAgent = UserAgent::random([
  'os_type'     => 'Windows',
  'device_type' => 'Desktop',
]);

$proxy = new Proxy(PROXY_HOST, PROXY_PORT, PROXY_USER, PROXY_PASS);

$browser = new Browser($client, $userAgent, SEARCH_LANGUAGE);
$browser->setProxy($proxy);

$googleUrl = new GoogleUrl(SEARCH_GOOGLE_HOST);
$googleUrl->setSearchTerm(SEARCH_KEYWORD);

$googleClient = new GoogleClient($browser);
// Set a custom header.
$googleClient->setHeader('Referer', 'https://www.google.hu/');
$response = $googleClient->query($googleUrl);
```

We can set a custom header with `$client->setHeader()` too:
```php
$client->setHeader('Referer', 'https://www.google.hu/');
```

What do you think? :)

**Comment by gsouf** [comment]
@lonalore That's not the best solution because the do double work on adapters and on google client.

I was a bit short in details in my previous message.

http client are actually working with psr7 request that already support headers, for instance with curl client headers are already taken from the given request: https://github.com/serp-spider/http-client-curl/blob/v0.2.0/src/CurlClient/Curl.php#L166
I think that your changes on the http interface are a bit of a duplicate with what is done here.

Concerning the changes you made on the google client that's a back to what was done in version 0.1. In version 0.1 the google client was responsible for managing the request by itself, but it was confusing and not very good for reusability. That's why in version 0.2 we introduced the browser class that is aimed to handle the logic of the request.

To be more precise:

- Google client is aimed to stick together a browser, an url, and a google response with some logic
- Http client are aimed to send given psr-7 requests, without opinion
- The browser is here to create the request, customize it if needed and send it through http client

Given this logic only the browser should be responsible for managing the requests (and that's why I introduced it in version 0.2). [See the abstract browser class](https://github.com/serp-spider/core/blob/master/src/Core/Browser/AbstractBrowser.php#L39), there is already a method to set default headers on all requests of the browser, and this method might be changed to be augmentable.


Well, to come back to your changes, I though a bit more and I'm not sure it's a good thing to add this logic on the http clients. In my opinion http clients should remain stateless.


Now that was said, in your case what you want to do is to set a referrer for google searches and this kind of thing should be done only when you are searching google, right ? In this case only requests sent by google client should have this referrer instead of all request sent by the browser. If this is true I was thinking of an exception to what I just said. Maybe should we give the ability to google client to modify the request before it's sent thanks to a callback. For instance that would be:

```php
$googleClient->setRequestHandler(function(RequestInterface $request){
    $request = $request->withHeader('referrer', 'mycustomreferer');
    return $request;
});
```

The workflow would be: 

- google client asks browser to create a request
- google client modifies the request with the given handler
- google client asks the browser to send the request

And in this situation most of the request logic remains in the browser (because it's its job) and google specific features are in the google client. To do this the [navigateToUrl](https://github.com/serp-spider/search-engine-google/blob/v0.2.0/src/GoogleClient.php#L66) must be replaced with ``$browser->prepareRequest  + givenHandler + $browser->sendRequest``

Considering my answer, what's you opinion on that?

**Comment by gsouf** [comment]
PS: Though I'm not sure it's good to set headers in a http client, I'm still thinking of it. If you see use cases where it can be profitable feel free to let me know, but at the moment I dont see any because all the tools that are needed are already available in browser.

**Comment by gsouf** [comment]
Phantomjs customheader implemented in #5.

Documentation TODO
